File: benchmarks/kdtree-raycast/lang_cpp/sampling.cpp
#include "vector.h"

#include <cassert>
#include <cmath>

const float PI = 3.14159265358979323846f;

Vector uniformSampleSphere(double u1, double u2)
{
    assert(u1 >= 0.0 && u1 < 1.0);
    assert(u2 >= 0.0 && u2 < 1.0);

    double z = 1.0 - 2.0 * u1;
    double r = sqrt(1.0 - z*z);
    double phi = 2.0 * PI * u2;
    double x = r * cos(phi);
    double y = r * sin(phi);
    return Vector(x, y, z);
}

File: benchmarks/kdtree-raycast/lang_go/sampling.go
package main

import "math"

func UniformSampleSphere(u1, u2 float64) Vector64 {
	z := 1.0 - 2.0 * u1
	r := math.Sqrt(1.0 - z*z)
	phi := 2.0 * math.Pi * u2
	x := r * math.Cos(phi)
	y := r * math.Sin(phi)
	return Vector64{x, y, z}
}
